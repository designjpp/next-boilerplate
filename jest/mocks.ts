// example mock data that will replace API data
export const mockedData = [
	{
		id: 1,
		name: 'one',
		description: 'one',
	},
	{
		id: 2,
		name: 'two',
		description: 'two',
	},
];
