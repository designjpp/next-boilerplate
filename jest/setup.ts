/* eslint-disable @typescript-eslint/no-empty-function */
import '@testing-library/jest-dom/extend-expect';
import { loadEnvConfig } from '@next/env';
import { toHaveNoViolations } from 'jest-axe';
import jestFetchMock from 'jest-fetch-mock';

import { server } from './server';

expect.extend(toHaveNoViolations);

// Establish API mocking before all tests.
beforeAll(() => {
	const projectDir = process.cwd();
	loadEnvConfig(projectDir);

	server.listen();
});

// Reset any request handlers that we may add during the tests,
// so they don't affect other tests.
afterEach(() => server.resetHandlers());
// Clean up after the tests are finished.
afterAll(() => server.close());

window.matchMedia =
	window.matchMedia ||
	function () {
		return {
			media: '',
			matches: false,
			onchange: function () {},
			dispatchEvent: function () {
				return false;
			},
			addListener: function () {},
			removeListener: function () {},
			addEventListener: function () {},
			removeEventListener: function () {},
		};
	};

jestFetchMock.enableMocks();
