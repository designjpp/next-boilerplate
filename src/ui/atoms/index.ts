export * from './button';
export * from './mouse';
export * from './scrollbar';
export * from './wow-title';
