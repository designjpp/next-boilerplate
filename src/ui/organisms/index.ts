// Put your organisms here
export const Organism = () => null;
