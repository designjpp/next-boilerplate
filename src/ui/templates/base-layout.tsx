import Link from "next/link";

import type { FC } from "react";

export const GBaseLayout: FC = ({ children }) => (
  <>
    <nav>
      <Link href="/">
        <a>Go to home</a>
      </Link>
      <Link href="/redux-sample">
        <a>Go to /redux-sample</a>
      </Link>
      <Link href="/redux-sample/server-side-props">
        <a>Go to /redux-sample/server-side-props</a>
      </Link>
    </nav>
    {children}

    <style jsx global>{`
      *:not(i) {
        font-family: Inter sans-serif;
      }

      #__next {
        @apply relative flex flex-col min-h-screen m-0 overflow-hidden text-base max-w-screen;

        padding: env(safe-area-inset-top) env(safe-area-inset-right)
          env(safe-area-inset-bottom) env(safe-area-inset-left);
      }
    `}</style>

    <style jsx>{`
      nav {
        @apply p-3 text-white bg-black;

        a {
          @apply p-3 mr-6;
        }
      }
    `}</style>
  </>
);
