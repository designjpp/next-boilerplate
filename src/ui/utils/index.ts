export * from './media-queries';
export * from './tailwind-config';
