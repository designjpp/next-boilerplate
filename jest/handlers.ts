import { rest } from 'msw';

import { mockedData } from './mocks';

// local server example only - to test that 'an' API call will be intercepted
export const handlers = [
	rest.get('http://localhost:3004/result', (req, res, ctx) => {
		return res(ctx.json(mockedData));
	}),
];
