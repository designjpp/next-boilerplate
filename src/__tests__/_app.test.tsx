import { render } from '@testing-library/react';
import { axe } from 'jest-axe';
import { create } from 'react-test-renderer';

import { mockUseRouter } from '__mocks__/mock-router.jest';
import App, { reportWebVitals } from 'pages/_app';

import { BaseLayout } from 'ui/templates';

import type { NextPage, NextPageWithLayout } from 'next';
import type { NextWebVitalsMetric } from 'next/app';
import type { Router } from 'next/router';

const router = mockUseRouter();

describe('_app.ts', () => {
	it('should be accessible', async () => {
		const { container } = render(
			<App Component={() => null} pageProps={{}} router={router as Router} />
		);
		const results = await axe(container);
		expect(results).toHaveNoViolations();
	});

	it('should match snapshot', () => {
		const component = create(
			<App Component={() => null} pageProps={{}} router={router as Router} />
		).toJSON();
		expect(component).toMatchSnapshot();
	});

	it('should match snapshot while an error occurred', () => {
		const component = create(
			<App
				Component={() => {
					throw Error('error');
				}}
				pageProps={{}}
				router={router as Router}
			/>
		).toJSON();
		expect(component).toMatchSnapshot('error');
	});

	it('should match snapshot while a layout is attached to page', () => {
		const Component: NextPageWithLayout = () => null;
		Component.Layout = {
			Component: BaseLayout,
			props: {},
		};

		const component = create(
			<App
				Component={Component as NextPage}
				pageProps={{}}
				router={router as Router}
			/>
		).toJSON();
		expect(component).toMatchSnapshot('with-layout');
	});

	it('should match snapshot while a layout is attached to page and the props of layout is customized', () => {
		const Component: NextPageWithLayout<
			{ layoutClassName: string },
			{ className?: string }
		> = () => null;
		Component.Layout = {
			Component: BaseLayout,
			props: ({ layoutClassName }) => ({ className: layoutClassName }),
		};

		const component = create(
			<App
				Component={Component as NextPage}
				pageProps={{ layoutClassName: 'text-white' }}
				router={router as Router}
			/>
		).toJSON();
		expect(component).toMatchSnapshot('with-layout-customized-props');
	});
});

describe('reportWebVitals', () => {
	const ORIGINAL_ENV = process.env;
	afterAll(() => {
		process.env = ORIGINAL_ENV;
	});

	let consoleLog: jest.SpyInstance;
	const webVital: NextWebVitalsMetric = {
		id: '1',
		startTime: 1,
		value: 1,
		label: 'web-vital',
		name: 'FCP',
	};

	beforeEach(() => {
		jest.resetModules();

		process.env = {
			...ORIGINAL_ENV,
			NODE_ENV: 'test',
			NEXT_PUBLIC_WEB_VITALS: 'true',
		};

		consoleLog = jest.spyOn(console, 'log').mockImplementation(() => undefined);
	});

	it('should log web vitals when the environment variable is available', () => {
		reportWebVitals(webVital);
		expect(consoleLog).toHaveBeenCalledWith(webVital);
	});

	it('should not log web vitals when the environment variable is unavailable', () => {
		process.env = {
			...ORIGINAL_ENV,
			NODE_ENV: 'test',
			NEXT_PUBLIC_WEB_VITALS: '',
		};

		reportWebVitals(webVital);
		expect(consoleLog).not.toHaveBeenCalled();
	});
});
