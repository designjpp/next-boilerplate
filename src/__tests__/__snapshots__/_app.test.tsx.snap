// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`_app.tsx Analytic integration should match snapshot while noIndex is false: noIndex-is-false 1`] = `null`;

exports[`_app.tsx Analytic integration should match snapshot while noIndex is true: noIndex-is-true 1`] = `null`;

exports[`_app.tsx Layout per page should match snapshot while a layout is attached to page and the props of layout is customized: with-layout-customized-props 1`] = `
<main
  className="text-white"
/>
`;

exports[`_app.tsx Layout per page should match snapshot while a layout is attached to page: with-layout 1`] = `<main />`;

exports[`_app.tsx should match snapshot 1`] = `null`;

exports[`_app.tsx should match snapshot while an error occurred: error 1`] = `
<p>
  An error occurred on client
</p>
`;
